### ISSUE TITLE
Rolling shutter compensation?

### ISSUE THREAD

**Issue by AndyLeeRobinson** [georgmartius/vid.stab#63]
Hi, I've used the deshaker plugin with VirdualDub for many years, and it can do a fairly good job of reducing rolling shutter distortion, both in lateral and vertical motion, but I don't see this ability with this plugin.
If the camera moves laterally, the image is skewed, and if moved vertically then the image is either squashed or stretched. This also depends on each particular camera used, so would need to analyse markers on the corners of the frames to try to determine the amount of compensation required.
Needless to say, this is complicated and I don't have the skills to be able to contribute!
Even rolling shutter compensation can't remove higher frequency distortions such as jello from multicopter footage.
If someone can crack that using a matrix to morph-compensate each frame for jello removal, they'd be a superstar!

**Comment by adrium** [comment]
That would be a very nice feature! For sunny outdoor scenes, the rolling shutter effect is less extreme. But otherwise, most other videos look distorted after stabilized...

btw, there is a nice paper about it using a neural network:
https://apvijay.github.io/pdf/2017_cvpr.pdf

**Comment by georgmartius** [comment]
Cool. Sounds like a good idea.

**Comment by andrixnet** [comment]
I'm not sure how it is implemented, but here is some usage documentation explaining about it for the "deshaker" filter for VirtualDub: http://www.guthspot.se/video/deshaker.htm

**Comment by karbiv** [comment]
Almost impossible to do it right. Such algorithms were developed for commercial software when industry started transitions from CCD image sensors to CMOS.
CMOS were worse in most aspects, but they are easier to produce. CCD included on-chip(on-sensor) memory to store all pixels at once - global shutter, no distortion.

Only recent(GoPro 7) CMOS sensors are fast enough to make rolling shutter artifacts unnoticeable. Or they calibrate in-built rolling shutter compensation. GoPro 6-7-8 use gyroscope to stabilize footage in real time, that info can also be used to fix rolling shutter.

**Comment by andrixnet** [comment]
I've worked with VirtualDub + deshaker a number of times and got some pretty good results. Since the rolling amount is guesswork for my devices, I did several trial-n-error runs to get the best footage.

**Comment by karbiv** [comment]
Yeah, I deleted my prev comment. I explore libvidstab code and saw some data prepared for "rolling shutter" compensation: transformtype.h; VSTransform structure.
But I work with differing workflow - taking fisheye video(wide angle) and stabilizing it. It can be passed to libvidstab as is and it somewhat works, because libvidstab is lens projection agnostic. Also each frame can be transformed to rectilinear(with Hugin) before libvidstab, and I'm looking for possibilities to make libvidstab aware of rectilinear lens projection("standard").

Rolling shutter is severe in some cameras not meant by engineers for videos, but marketed as video ready. Almost all smartphone sensors introduce severe rolling shutter to the raw video.
It can be fixed by an algorithm that gets as input a data about camera movements.
Smartphones get that data from in-built gyroscope.

Libvidstab has to analyze frames already distorted by rolling shutter to get that approximation of camera movements to fix that rolling shutter. From the inception such algorithm will be of lower quality than any built-in stabilization of modern cameras like GoPro 7-8 and smartphones that have gyroscopes. Well, it's possible but not interesting for some perfectionist programmer.

But modern CMOS cameras are already free of rolling shutter. $100 like Runcam 5 that is for sport/enthusiast quadcopters, uses one of the latest Sony image sensors.
Is it reasonable to make a rolling shutter compensation for those who buy $400~$700 smartphones with crappy cameras for video? Those cameras in normal mode would have it's own rolling shutter compensation. Just look on youtube what GoPro 7-8 does with electronic stabilization, better than any optical.

**Comment by karbiv** [comment]
VirtualDub was unique in some way, it was one of the first open source video editors even before the proliferation of usable Linux distros.
It's unique because it managed to create a rolling shutter compensation for not commercial product.
Deshaker plugin for VirtualDub by Gunnar Thalin is a donationware, for Windows users. There's a list of camera models, settings for them. Maintaining and polishing something similar for a Linux library is problematic.
Most programmers lack any motivation to work with rolling shutter when they understand why it exists in the first place. It's probably not an engineering problem, but a strategy created by marketing departments.

CCD sensors didn't have that "rolling shutter" artifact. All pixels of a frame were read in one moment, into an on-sensor memory buffer as I understand. No distortions, no jello caused by vibration, global shutter.
CMOS sensors are cheaper to produce on a standard line/facility. Simpler physical interface of a chip that allows reading one line of pixels at a moment of time. So it must be coupled with a special CPU, like Sony's Exmor line, that is programmed to understand particular image sensor's characteristics.

An eye opening video(was for me) from HongKong 2011 that demonstrated the state of CCD and CMOS technologies by that time: https://www.youtube.com/watch?v=q8OtTzOpZcE
CCD was perfect for videos. In comments users ask which one is CCD, because they came from many articles that claimed how good new CMOS cameras are and can't believe the reality.

Generic rolling shutter compensation in software can't be ideal. I think that users of commercial software DEMANDED the fix for it. It's difficult to explain to non-techy users that their camera is crappy for videos, because corporations didn't think about videos, or marketing departments deliberately made it suck for videos.

**Comment by karbiv** [comment]
I have to agree with other commenters about annoyance of RS, after some thinking.
Rolling Shutter as a visible effect in videos is probably won't go away eventually, because it probably is balanced with light sensitivity of image sensors, just a guess.
It's noticeable in Sony and GoPro competition, Sony cams are better in low light but introduce more rolling shutter than GoPro.

Rolling Shutter is the last obstacle in achieving almost ideal software stabilization of wide angle(not only) lenses. Ideal stabilization can be done only with other gem of free software: Hugin Panorama Stitcher, command line tools that come with it.
Hugin provides unmatched lens calibration(nothing to compare with), distortion correction, **and correct image transformation after detect pas of libvidstab**.
Libvidstab using intensity based detection works as good as feature based detection of Hugin, if not better. Feature based detection is very slow.

Visible rolling shutter of a stabilized video from action cam, in rectilinear projection, original is fisheye:

https://www.youtube.com/watch?v=d9fpwFl_yFU

That's a challenge...

**Comment by karbiv** [comment]
Libvidstab internally computes from _localmotion_ rectangles a mean vector of movement(direction and amplitude of frame shift) and frame rotation mean angle. Those two parameters for each frame are enough data for rolling shutter compensation.

It may require additional optional pass. Detect phase analyzes distorted frames of original video, so it may be useful to repeat detection on frames with fixed rolling shutter, when distortion due to it was severe.
But all that nitpicking would be more obvious in resulting video if libvidstab could correct frame pixel positions during transform pass, i.e. _lens projection correction_. Currently in libvidstab result is good for narrow field of view videos, when distance from optical center is not significant, and that's how libvidstab is used mostly.

A demo of libvidstab detect phase for a wide field of view video, with transform pass done in Hugin:

Original: https://www.youtube.com/watch?v=e6ayclySnzo
Stabilized and reprojected: https://www.youtube.com/watch?v=l6AFy2YCiYc

Stabilized video is in _cylindrical projection_(in Hugin), when vertical lines are straight. Horizontal FOV is 137 degrees.
This was shot on a limit of RS revealing itself. When not shot accurately, buildings can stretch vertically on each step. In smartphones RS is more pronounced.

**Comment by karbiv** [comment]
Main conclusion - **it's a comparatively computationally expensive per-pixel operation with interpolation**.
The scanning of lines in a video can be _top-bottom_, and _bottom-up_ - because smartphones can be held in different orientations and action cameras have a FLIP mode. So in effect the scanning of lines in a video can be vertical, left-to-right and right-to-left - **all cases should be specified by configuration parameters**.

AndyLeeRobinson:
> If the camera moves laterally, the image is skewed, and if moved vertically then the image is either squashed or stretched. This also depends on each particular camera used, so would need to analyse markers on the corners of the frames to try to determine the amount of compensation required.

Also during fast camera rotation around optical axis(roll) left part is stretched vertically and right part compressed vertically, during lines scanning top-bottom or bottom-up.(lines of pixels in an image sensor are scanned in one direction, but filmed objects on side of the center move in opposite directions during camera fast rolls)

adrium
> For sunny outdoor scenes, the rolling shutter effect is less extreme. But otherwise, most other videos look distorted after stabilized...

That's introduced mostly by _motion blur_ of pixels, probably some kind of varying ISO sensitivity - image sensor's pixels must be exposed longer to less light.
It's impossible to compensate using, but new cameras increase their ISO sensitivity.

It's easier to develop compensation algorithm in something like **scikit-image** using Python, than eventually it can be written in pure C for FFMPEG.

How its usage would look in practice:
This algorithm should detect shakiness of a video as usual, then produce a video without stabilization but with corrected rolling shutter warping of all kinds(horizontal, vertical, roll), and that video should be an input of another stabilization detect pass for better precision.

**Comment by karbiv** [comment]
> If the camera moves laterally, the image is skewed, and if moved vertically then the image is either squashed or stretched.

I successfully tested Rolling Shutter correction using libvidstab and scikit-image library(python).
There's also rolling shutter distortion from camera rotation, one side of the frame is squashed, other is stretched. When camera is rotated, first line is normal, each subsequent line is rotated more around the center.

Rolling Shutter correction parameter should be one number.(I used two, a separate one for rotation because I still need to realize how to deduce it from the main parameter).

Other important toggle parameter is "lines scan direction": top-down or bottom. depends on how camera is held. Action cams have a "flip" option that turns image upside down.

**Comment by georgmartius** [comment]
Closing this: **rolling shutter compensation is not something we plan to implement.** Better to say so plainly than leave it open for another eight years.

The reason is structural rather than a matter of effort. vid.stab's entire model is *one transform per frame*: detection produces a set of local motions, those are fitted to a single similarity transform (translation, rotation, uniform zoom), and the second pass warps the frame by it. Rolling shutter is by construction not that — each scanline is exposed at a different instant, so correcting it needs a motion estimate *per scanline*, interpolated along the frame, plus a model of the camera's angular velocity during the readout interval. That is a different estimator, a different file format, and a different warp; it is not a parameter that could be added to the existing pipeline.

Doing it properly also needs information vid.stab does not have and cannot recover from the pixels alone — readout time and direction, which vary per sensor and per mode — and generally the lens model too, since rolling-shutter correction and lens distortion interact. The literature to start from is Forssén & Ringaby, *"Rectifying rolling shutter video from hand-held devices"* (CVPR 2010).

In practice, if your footage suffers from it, the tools that handle it well do so because they were built around a per-scanline model from the start — Gyroflow if you have gyro data from the camera, or the vendor's own stabilization. Those will beat anything that could reasonably be bolted on here.

Anyone who wants to attempt it is genuinely welcome to, but it should be treated as a new estimator rather than a feature request against this one, and the project is currently looking for a maintainer rather than taking on work of this size.

**Comment by andrixnet** [comment]
By "new estimator" do you mean a different filter for ffmpeg or a different subcomponent of vid.stab selectable via filter options?

**Comment by karbiv** [comment]
Yes, Rolling Shutter Correction is not easily generalizable, it should be implemented for a particular lens/sensor combination, like Raspberry Pi cameras. Users usually lack knowledge about camera's hardware.

Some cameras can be mounted upside-down with an option in menu, but when looking at old video archive from that camera it's not known what mode was used - was it top-bottom or bottom-up scan in image sensor.
Some cameras have Optical Image Stabilization to reduce motion blur, they remove tremor and videos are more suitable for software Rolling Shutter correction (with some caveats of lens perspective), but the data from gyroscope is only partially valid after OIS.

Inertial Measurement Unit(gyroscope) is the only precise method to correct Rolling Shutter. With sample rate of 0.8-1.6 kHz it collects movement data for each block of 40-80 scanlines at 1080p 60FPS.
All these corrections are per pixel and require expensive chips.
Modern system-on-a-chip like Raspberry Pi Zero 2 has capabilities to operate on HD frames, so it's possible to create a camera with corrections comparable to those in GoPro, DJI.

For software generalized implementation the resulting program won't be very useful or effective.

**Comment by andrixnet** [comment]
@karbiv I can only talk from end-user experience, with VirtualDub and Deshaker plugin. This one has a rolling shutter compensation parameter (shown as a % percentage). It can make quite the big difference. 
Given the plugins documentation and instructions, I did measurements with several of my cameras under several circumstances and the resulting parameter from those measurements really help in the compensation. 
So yes, a generalized implementation is possible, with quite good results, as per my experience. _Although I know nothing about it's inner workings._ 

A scriptable solution such as ffmpeg with capability to stabilize & rolling shutter compensation would be great. 
Till then, if ever, Deshaker remainsa very good alternative, albeit with per video file manual editing.

There are differences between behaviour with OIS on or off, also different recording resolutions and zoom levels _(considering only optical zoom)_ 
Orientation sensor I've seen so far being ignored by ffmpeg. If such metadata gets taken into account, surely there must be a switch to ignore orientation metadata and then the proper raw footage is processed and a simple rotation can be made afterwards.

**Comment by karbiv** [comment]
@andrixnet Deshaker introduced parameterized(%) Rolling Shutter correction in 2008. Both Deshaker and libvidstab operate in 2D planar space. It works for narrow Field of View cameras.

"2D planar space" means that frames are just moved up/down/left/right as is, without recalculating each pixel. In wide Field of View cameras, objects in corners of a frame look bigger, for example feet look bigger. That's how a normal rectilinear lens works - all straight lines in real world look straight in videos, this rule makes all objects stretched the farther they are from the center of the lens. When such frames are deshaked, pixels in corners could be moved closer to the center without recalculating lens projection corresponding to their new position in a frame.

To convert it to a 3D space calculation, parts of libvidstab must be rewritten. In Pass 1, analysis of pixel movements must be converted to the angle of rotation around camera. For instance, 10 pixels movement in the center of the frame is not the same as 10 pixels movement in the corner of the frame - angles of movement around camera are different.

To use such model a user must specify as option **Horizontal Angle of View** of the camera. Having this parameter the algorithm knows exact angle between any 2 pixels in the frame. Now frames should be not just shifted but **reprojected**.

That reprojection is required for any pixel translations - Rolling Shutter correction.

With fish-eye lenses _Angle of View_ parameter is not enough. There are lenses that current FFMPEG plugins and software like Blender can't correct to rectilinear, like Zeiss lens in Sony action cams with 170° angle of view. Only Hugin software is able to convert that image to rectilinear by using GUI. Polynomial(3-4 numbers) parameters that are used to correct fish-eye in software is not intuitive and it doesn't work for some lenses.
The solution is to use a sequence of numbers - coordinates on a 2d graph representing dots on a curve of a lens(from center to corners along radius). The program can interpolate those numbers into a precise graph with thousands of dots. User would add those number until it outputs a rectilinear frame where all lines are straight, thus discovering almost exact graph for any lens.

This data about lens curve should be enough to reproject pixel translation for any lens when doing Rolling Shutter correction. In addition to more precise movement analysis and no new distortions in corrected frames.

**Comment by adrium** [comment]
For completeness sake: Deshaker was open-sourced
https://github.com/gu-t/Deshaker/blob/master/DeShaker.cpp

Interesting is the following comment:

> Rolling shutter video needs to calculate two sets of general motion parameters. One for the first line and one for the last. Everything in between is interpolated between these two.

It sounds like the implementation is rather simple, but effective.

**Comment by karbiv** [comment]
That implementation is TOO simple.
The best demonstration of pixel reprojection - Google Street View. Open any street, position some building's window or a car in the center of the screen, then move it towards one of the screen corners - it stretches and becomes bigger. That's how it should be when shifting any frame. Every pixel must be reprojected.

The same is with Rolling Shutter correction. Moving pixels of scanlines requires reprojection to avoid introduction of other distortions similar to those caused by Rolling Shutter.

Rolling Shutter Correction for a fish-eye lens requires the exact description of the lens in form of data table or something. Otherwise fish-eye video should be transformed to rectilinear in a separate pass by some FFMPEG plugin. The problem that those plugins are not always usable. And the resulting video will have much bigger resolution, some deliberate aspect ratio, then other transformation passes are needed like cropping, sharpening.

It's not that the solution is unknown - it's a lot of new code and testing.

**Comment by adrium** [comment]
@karbiv Thanks, your explanation is clear and understandable.

Deshaker was a popular tool and I remember it was recommended by numerous websites and forums. Apparently, it helped people.

Therefore, calling the Deshaker implementation 'TOO simple' is not really fair.

@georgmartius From what I understand, a similar transform should theoretically be possible with vid.stab's model of _one transform per frame,_ right? Comparing the difference of upper and lower scanlines in the first pass and a shear transform in the second.

**Comment by andrixnet** [comment]
@adrium thank you for your support. 

Based on the above information, from end-user perspective, it should be possible without too much complexity to implement a switchable and parametric rolling shutter compensation based on Deshaker, more so as it seems to be open-source, as a basic form of compensation and with the documented caveat as to what it's capabilities and limitations are, as are many other known ffmpeg features. 

That would bring ffmpeg scriptable stabilization capability comparable to that of deshaker, which would be way better then no compensation for the want of a super-mega-complex implementation that gives mostly the same results except for certain camera contexts where it would perform better with the right parameters (which merit a study on the level of Michael Koch's book).

Most would prefer a regular car that takes you from A to B rather then wait for a hypotethical future Ferrari.

As for my part, I could contribute by end-user testing, that is processing the same video with both deshaker and vid.stab, using my known measurements for my cameras and compare the results and discuss the influence of various parameters.

**Comment by georgmartius** [comment]
Thanks for the comments. We are soon having already FoV correction and lens correction. Maybe I will get the rolling shutter thing as well. let's see.

**Comment by karbiv** [comment]
> We are soon having already FoV correction and lens correction. Maybe I will get the rolling shutter thing as well. let's see.

In this case, Rolling Shutter correction only for rectilinear cameras  https://github.com/georgmartius/vid.stab/issues/139#issuecomment-5231839696.

Lens representation is flawed in software. Some fish-eye lenses can be rectified only by Hugin, nothing else is capable.

The only universal type of parameter for lenses would be a graph - 7-10 normalized numbers(0<1.0) representing dots of a graph and interpolated by Bezier curves or something.
It leads me once again to a custom implementation.

**Comment by mcprat** [comment]
> > We are soon having already FoV correction and lens correction. Maybe I will get the rolling shutter thing as well. let's see.

> It leads me once again to a custom implementation.

yeah perhaps there could be custom profiles for this or index values to adjust

the situation reminds me of calibration of reference microphones

**Comment by andrixnet** [comment]
As I said before, I am sure many users will be more then happy with a basic compensation, with few parameters to fiddle with until the results are satisfactory.

Also, perhaps at a later date, more advanced methods of compensation could be added, with more complex parametrization, which will benefit those with more specific use cases for which the additional complexity solves their context.

Like the example above, a "good enough" for most microphones in the near future is still better then nothing for a long time tome come waiting for a "super" for reference microphones.

Thank you.
